src/EventListener/ExceptionListener.php

```diff
@@ -0,0 +1,416 @@
+     *
+     * @param Response $response
+     *
+     * @return Response
+     */
+    private function setXStatusCode(Response $response)
+    {
+        if (!$response->headers->has('X-Status-Code')) {
+            $response->headers->set('X-Status-Code', $response->getStatusCode());
+        }
+
+        return $response;
+    }
+
+    /**
+     * Check if the exception or any exception in the chain implements the response exception interface.
+     *
+     * @param \Exception $exception The exception to walk on.
+     *
+     * @return null|Response
+     */
+    private function checkResponseException($exception)
+    {
+        do {
+            if ($exception instanceof ResponseExceptionInterface) {
+                return $exception->getResponse();
+            }
+        } while (null !== ($exception = $exception->getPrevious()));
+
+        return null;
+    }
+
+    /**
+     * Check if the exception or any exception in the chain inherits from HttpException and is understood by us.
+     *
+     * @param \Exception $exception The exception to walk on.
+     *
+     * @param Request $request    The HTTP request.
+     *
+     * @return null|Response
+     */
+    private function checkHttpExceptions(\Exception $exception, $request)
+    {
+        do {
+            if ($response = $this->checkHttpException($exception, $request)) {
+                return $response;
+            }
+        } while (null !== ($exception = $exception->getPrevious()));
+
+        return null;
+    }
+
+    /**
+     * Check if the Contao 404 should be rendered.
+     *
+     * @param \Exception $exception The exception to test.
+     *
+     * @return bool
+     */
+    private function isRenderingOfContao404PossibleFor($exception)
+    {
+        // If not an 404 exception or the frontend has not been booted, we can not render.
+        if (!($exception instanceof NotFoundHttpException && defined('BE_USER_LOGGED_IN'))) {
+            return false;
+        }
+
+        // It is hopeless to render a 404 page when no root page or no page at all is published.
+        if ($exception instanceof RootNotFoundHttpException || $exception instanceof NoPagesFoundHttpException) {
+            return false;
+        }
+
+        // If no 404 page handler has been registered, we also can not render.
+        if (!(isset($GLOBALS['TL_PTY']['error_404']) && class_exists($GLOBALS['TL_PTY']['error_404']))) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Try to render a Contao 404 page first.
+     *
+     * @param \Exception $exception The exception to check.
+     *
+     * @return null|Response
+     *
+     * @throws \Exception When anything went really haywire during rendering of the page.
+     */
+    private function tryToRenderContao404($exception)
+    {
+        if (!$this->isRenderingOfContao404PossibleFor($exception)) {
+            return null;
+        }
+
+        static $processing;
```

Comment:
I think this should be a private static class variable (Symfony does that in `Symfony\Component\Config\Loader\FileLoader`).